background.js

```diff
@@ -1,27 +1,171 @@
 const pluginFlag = "pluginFlag"
-var shouldTranslate = false
-let initalRequestMethod = "shouldTranslate"
+let shouldTranslate = false
+let initialRequestMethod = "shouldTranslate"
 let updateRequestMethod = "updateShouldTranslate"
-let reloadRequestMethod = "reloadShouldTranslate"
+let translatedRequestMethod = "translated"
+const reTranslatedRequestMethod = "reTranslated"
+var currentTranslatedPage = [];
+var currentTabID = 0
+var previousTabID = 0
+const pageSwitchedRequestMethod = "pageSwitched"
+let refreshRequested = false
 
+const BrowserType = {
+    chrome: Symbol("chrome"),
+    safari: Symbol("safari"),
+    firefox: Symbol("firefox"),
+    unknown: Symbol("unknown")
+}
+
+setTimeout(() => {
+    chrome.storage.local.get(pluginFlag, (result) => {
+        shouldTranslate = result.pluginFlag || false
+        if (!shouldTranslate) {
+            chrome.action.setIcon({ path: { "128": "/source/intro/swiftLogo-closed.png"} }).then(r => {})
+        }
+    });
 
-chrome.storage.local.get(pluginFlag, (result) => {
-    shouldTranslate = result.pluginFlag || false
-  });
+}, 1)
 
 chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
     if (request.type === updateRequestMethod) {
         shouldTranslate = request.data;
-        chrome.tabs.query({ active: true, currentWindow: true }, function (tabs) {
-          chrome.tabs.reload(tabs[0].id);
+        currentTranslatedPage = []
+        currentTabID = 0
+        chrome.tabs.query({}, function (tabs) {
+           chrome.tabs.query({active: true, currentWindow: true}, function (tabs) {
+               const activeTab = tabs[0]
+               updateTabId(activeTab.id)
+               if (activeTab.url.includes("developer.apple.com")) {
+                   updateLogo(true)
+               } else {
+                   updateLogo(false)
+               }
+           })
+
+            refreshRequested = true
+
+            let running = []
+            for (let tab of tabs) {
+                if (tab.url.includes("developer.apple.com")) {
+                    running.push(chrome.tabs.reload(tab.id))
+                }
+            }
+
+            Promise.allSettled(running).then()
         });
-    } else if (request.type === initalRequestMethod) {
+        return true
+    } else if (request.type === initialRequestMethod) {
         sendResponse({shouldTranslate: shouldTranslate});
+    } else if (request.type === translatedRequestMethod) {
+        currentTranslatedPage.push(sender.tab.id)
+        updateLogo(true)
+        return true
     }
 });
 
 function delay(i) {
   setTimeout(() => {
-    console.log(i);
   }, i * 1000);
 }
+
+chrome.tabs.onUpdated.addListener(function(tabId, info, tab) {
+    if (refreshRequested) {
+        refreshRequested = false
+        return true
+    }
+
+    chrome.tabs.query({ active: true, currentWindow: true }, function (tabs) {
```

Comment:
We can unify the handle tab logic into one method, and then both onUpdated and onActivated can call the method.